SimG4Core/DD4hepGeometry/plugins/DD4hep_GeometryESProducer.cc

```diff
@@ -0,0 +1,221 @@
+}
+
+G4GeometryESProducer::~G4GeometryESProducer() 
+{
+  delete m_attach;
+  delete m_kernel; 
+}
+
+void G4GeometryESProducer::updateMagneticField(EventSetup const& es) {
+  if(m_pUseMagneticField) {
+    // setup the magnetic field
+    ESHandle<MagneticField> pMF;
+    es.get<IdealMagneticFieldRecord>().get(pMF);
+    const GlobalPoint g(0.,0.,0.);
+    LogInfo("G4GeometryESProducer") << "B-field(T) at (0,0,0)(cm): " << pMF->inTesla(g);
+
+    sim::FieldBuilder fieldBuilder(pMF.product(), m_pField);
+    CMSFieldManager* fieldManager = new CMSFieldManager();
+    G4TransportationManager * tM = G4TransportationManager::GetTransportationManager();
+    tM->SetFieldManager(fieldManager);
+    fieldBuilder.build( fieldManager, tM->GetPropagatorInField());
+    LogInfo("G4GeometryESProducer") << "Magentic field is built";
+  }
+}
+
+void G4GeometryESProducer::beginLuminosityBlock(edm::LuminosityBlock&, edm::EventSetup const&) 
+{
+  // mag field cannot be change in new lumi section - this is commented out
+  //     updateMagneticField( es );
+}
+
+void G4GeometryESProducer::beginRun(const edm::Run &, const edm::EventSetup&)
+{}
+
+void G4GeometryESProducer::endRun(const edm::Run &,const edm::EventSetup&)
+{}
+
+void G4GeometryESProducer::produce(edm::Event & e, const edm::EventSetup & es)
+{
+  if(!m_firstRun) return;
+  m_firstRun = false;
+  
+  LogInfo("G4GeometryESProducer") << "Producing G4 Geom";   
+
+  m_kernel = G4RunManagerKernel::GetRunManagerKernel();   
+  if(m_kernel == nullptr) m_kernel = new G4RunManagerKernel();
+  LogInfo("G4GeometryESProducer") << " G4GeometryESProducer initializing ";
+  // DDDWorld: get the DDCV from the ES and use it to build the World
+  ESTransientHandle<DDCompactView> pDD;
+  es.get<IdealGeometryRecord>().get(pDD);
+  //  m_detDesc->apply("DD4hep_XMLLoader", 1, &arg);
+  G4LogicalVolumeToDDLogicalPartMap map;
+  SensitiveDetectorCatalog catalog;
+  const DDDWorld * world = new DDDWorld(&(*pDD), map, catalog, false);
+  m_registry.dddWorldSignal_(world);
+  //FIXME:??? m_registry.watchDDDWorld(world);
+
+  updateMagneticField(es);
+  
+  if(m_pUseSensitiveDetectors) {
+    LogInfo("G4GeometryESProducer") << " instantiating sensitive detectors ";
+    // instantiate and attach the sensitive detectors
+    m_trackManager = unique_ptr<SimTrackManager>(new SimTrackManager);
+    if(m_attach == nullptr) m_attach = new AttachSD;
+    {
+      pair<vector<SensitiveTkDetector*>,
+	   vector<SensitiveCaloDetector*>> 
+	sensDets = m_attach->create((*pDD), catalog, m_p, m_trackManager.get(), m_registry);
+      
+      m_sensTkDets.swap(sensDets.first);
+      m_sensCaloDets.swap(sensDets.second);
+    }
+
+    LogInfo("G4GeometryESProducer") << " Sensitive Detector building finished; found " 
+				<< m_sensTkDets.size()
+				<< " Tk type Producers, and " << m_sensCaloDets.size() 
+				<< " Calo type producers ";
+  }
+
+  for(Producers::iterator itProd = m_producers.begin();itProd != m_producers.end();
```

Comment:
@ianna , what is expected from this loop?

Reply:
Yet another good question:-)